opm/simulators/linalg/bda/openclSolverBackend.cpp

```diff
@@ -0,0 +1,783 @@
+/*
+  Copyright 2020 Equinor ASA
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <config.h>
+#include <cmath>
+#include <sstream>
+
+#include <opm/common/OpmLog/OpmLog.hpp>
+#include <opm/common/ErrorMacros.hpp>
+
+
+#define __CL_ENABLE_EXCEPTIONS
+#include <CL/cl.hpp>
+#include <opm/simulators/linalg/bda/openclKernels.hpp>
+
+
+#include <opm/simulators/linalg/bda/openclSolverBackend.hpp>
+#include <opm/simulators/linalg/bda/BdaResult.hpp>
+#include <opm/simulators/linalg/bda/Reorder.hpp>
+
+
+// iff true, the nonzeroes of the matrix are copied row-by-row into a contiguous, pinned memory array, then a single GPU memcpy is done
+// otherwise, the nonzeroes of the matrix are assumed to be in a contiguous array, and a single GPU memcpy is enough
+#define COPY_ROW_BY_ROW 0
+
+// Level Scheduling respects the depencies in the original matrix
+// Graph Coloring is more aggresive and is likely to change the number of linearizations and linear iterations to converge, but can still be faster on GPU because it results in more parallelism
+#define LEVEL_SCHEDULING 1
+#define GRAPH_COLORING   0
```

Comment:
Do these need to be macros? Are both tested?

Reply:
Both converge for NORNE_1A_MSW. GRAPH_COLORING is even faster, because it extracts more parallelism. But it does have more linear iterations then LEVEL_SCHEDULING. Their results have been visually verified with ResInsight.
However, only for the May commits. The most recent commits have a problem with converging.

Reply:
So the convergence problems are only for LEVEL_SCHEDULING?

Reply:
No, for the May versions of the opm repos, both LEVEL_SCHEDULING and GRAPH_COLORING finish NORNE_1A_MSW (see edited startpost). GRAPH_COLORING is faster because it extracts more parallelism, despite needed more linear iterations due to disregarding some row dependencies.
For the most recent opm repos versions, NORNE_1A_MSW does not converge with Dune, cusparse or opencl (LS) in reasonable time. I'm not sure if I tested opencl GC with these newest commits.

Reply:
I think runtime matters more than no of iterations. maybe change the default to the faster option and leave the rest for the future?

Reply:
Implemented in 5aa8dda